source/fe/mapping_q_generic.cc

```diff
@@ -2577,46 +2668,70 @@ MappingQGeneric<dim, spacedim>::transform_real_to_unit_cell(
-  else
     {
-      // for the MappingQEulerian type classes, we want to still call the cell
-      // iterator's affine approximation. do so by creating a dummy
-      // triangulation with just the first vertices.
-      //
-      // we do this by first getting all support points, then
-      // throwing away all but the vertices, and finally calling
-      // the same function as above
-      std::vector<Point<spacedim>> a =
-        this->compute_mapping_support_points(cell);
-      a.resize(GeometryInfo<dim>::vertices_per_cell);
-      std::vector<CellData<dim>> cells(1);
-      for (const unsigned int i : GeometryInfo<dim>::vertex_indices())
-        cells[0].vertices[i] = i;
-      Triangulation<dim, spacedim> tria;
-      tria.create_triangulation(a, cells, SubCellData());
-      initial_p_unit =
-        tria.begin_active()->real_to_unit_cell_affine_approximation(p);
+      initial_p_unit = cell->real_to_unit_cell_affine_approximation(p);
+      // in 1d with spacedim > 1 the affine approximation is exact
+      if (dim == 1 && polynomial_degree == 1)
+        return initial_p_unit;
     }
-  // in 1d with spacedim > 1 the affine approximation is exact
-  if (dim == 1 && polynomial_degree == 1)
+  else
     {
-      return initial_p_unit;
+      // else, we simply use the mid point
+      for (unsigned int d = 0; d < dim; ++d)
+        initial_p_unit[d] = 0.5;
     }
-  else
+
+  // in case the function above should have given us something back that lies
+  // outside the unit cell, then project it back into the reference cell in
+  // hopes that this gives a better starting point to the following iteration
+  initial_p_unit = GeometryInfo<dim>::project_to_unit_cell(initial_p_unit);
+
+  // perform the Newton iteration and return the result. note that this
+  // statement may throw an exception, which we simply pass up to the caller
+  const Point<dim> p_unit =
+    this->transform_real_to_unit_cell_internal(cell, p, initial_p_unit);
+  if (p_unit[0] == std::numeric_limits<double>::infinity())
+    AssertThrow(false,
+                (typename Mapping<dim, spacedim>::ExcTransformationFailed()));
+  return p_unit;
+}
+
+
+
+template <int dim, int spacedim>
+void
+MappingQGeneric<dim, spacedim>::transform_points_real_to_unit_cell(
+  const typename Triangulation<dim, spacedim>::cell_iterator &cell,
+  const ArrayView<const Point<spacedim>> &                    real_points,
+  ArrayView<Point<dim>> &                                     unit_points) const
+{
+  AssertDimension(real_points.size(), unit_points.size());
+  const std::vector<Point<spacedim>> support_points =
+    this->compute_mapping_support_points(cell);
+  const std::pair<DerivativeForm<1, dim, spacedim>, Tensor<1, spacedim>>
+    affine_factors = GridTools::affine_cell_approximation<dim>(
+      ArrayView<const Point<spacedim>>(support_points.data(),
+                                       GeometryInfo<dim>::vertices_per_cell));
```

Comment:
```suggestion
      ArrayView<const Point<spacedim>>(support_points));
```

Reply:
I see that this needs a comment. I intentionally only pick the first `2^dim` points to construct a linear mapping and ignore the rest. After all, this only serves for the initial guess.